core/client/src/main/java/alluxio/client/file/FileInStream.java

```diff
@@ -508,4 +515,82 @@ private void updateBlockInStream(long blockId) throws IOException {
       mShouldCacheCurrentBlock = mAlluxioStorageType.isStore();
     }
   }
+
+  /**
+   * Seeks to a file position. Blocks are not cached unless they are fully read. This is only called
+   * by {@link FileInStream#seek}.
+   *
+   * @param pos The position to seek to. It is guaranteed to be valid (pos >= 0 && pos != mPos &&
+   *            pos <= mFileLength)
+   * @throws IOException if the seek fails due to an error accessing the stream at the position
+   */
+  private void seekInternal(long pos) throws IOException {
+    seekBlockInStream(pos);
+    checkAndAdvanceBlockInStream();
+    mCurrentBlockInStream.seek(mPos % mBlockSize);
+  }
+
+  /**
+   * Seeks to a file position. Blocks are cached even if they are not fully read. This is only
+   * called by {@link FileInStream#seek}.
+   * Invariant: if the current block is to be cached, [0, mPos) should have been cached already.
+   *
+   * @param pos The position to seek to. It is guaranteed to be valid (pos >= 0 && pos != mPos &&
+   *            pos <= mFileLength).
+   * @throws IOException if the seek fails due to an error accessing the stream at the position
+   */
+  private void seekInternalWithCachingPartiallyReadBlock(long pos) throws IOException {
+    // Precompute this because mPos will be updated several times in this function.
+    boolean isInCurrentBlock = pos / mBlockSize == mPos / mBlockSize;
+
+    if (mShouldCacheCurrentBlock) {
+      // Cache till pos if seeking forward within the current block. Otheriwse cache the whole
+      // block.
+      readCurrentBlockToPos(pos > mPos ? pos : mFileLength);
+    }
+
+    // If seeks within the current block, directly seeks to pos if we are not yet there.
+    // If seeks outside the current block, seek to the beginning of that block first, then
+    // cache the prefix (pos % mBlockSize) of that block.
+    if (isInCurrentBlock) {
+      // We don't need to seek if we are at pos already. It won't work if you remove this early
+      // return because SeekBlockInStream closes the current cache stream.
+      if (mPos == pos) {
+        return;
+      }
+
```

Comment:
done. FYI, I am refactoring this file to make it clearer. I feel FileInStream is a little hard to follow and easy to make mistakes when I make changes to it.